### PR TITLE
Added method to list blobs and directories at the same time

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
This fixes the issue in #412, without asking users to rely on internal, undocumented methods. It also ensures that the Azure Storage SDK offers similar features to storage SDKs from AWS and others.

I could use some guidance on how to write tests for this, however, as understanding the test suite was no easy feat :)

### CODE REVIEW

**Comment by ItalyPaleAle** [comment]
@jiacfan any update here?

**Comment by ItalyPaleAle** [comment]
Thanks for the update!

In the meanwhile, I've fixed the typo you pointed out.

**Comment by jiacfan** [comment]
>         responseObject.listBlobsResult.entries.push(blobResult);

For BlobsAndDirectories type, would you add two container fileEntries and directoryEntries, which contains file and directory separately and keep alphabetical order?

---
Refers to: lib/services/blob/blobservice.core.js:5848 in ea8d8ac. [](commit_id = ea8d8acf768d3bbcf2b37ec4f59b981164a11d5c, deletion_comment = False)

File: lib/services/blob/blobservice.core.js [hunk 1]
```diff
@@ -1122,6 +1122,38 @@ BlobService.prototype.listBlobDirectoriesSegmentedWithPrefix = function (contain
...
+* @param {string}             [options.clientRequestId]           A string that represents the client request ID with a 1KB character limit.
+* @param {bool}               [options.useNagleAlgorithm]         Determines whether the Nagle algorithm is used; true to use the Nagle algorithm; otherwise, false.
+*                                                                 The default value is false.
+* @param {errorOrResult}      callback                            `error` will contain information
+*                                                                 if an error occurs; otherwise `result` will contain `entries` and `continuationToken`. 
+*                                                                 `entries`  gives a list of `[directories]{@link BlobResult}` and the `continuationToken` is used for the next listing operation.
+*                                                                 `response` will contain information related to this operation.
+*/
+BlobService.prototype.listBlobsOrBlobDirectoriesSegmentedWithPrefix = function (container, prefix, currentToken, optionsOrCallback, callback) {
```
Comment (jiacfan) [inline lib/services/blob/blobservice.core.js]: Would you add test for this new API as well?

**Comment by jiacfan** [comment]
@ItalyPaleAle 

Thanks for the quick response, that's great!

And we have some comments, it would be great if you can help to optimize it further, you know as rest api doc [list blobs](https://docs.microsoft.com/en-us/rest/api/storageservices/list-blobs), 

> Blobs are listed in alphabetical order in the response body, with upper-case letters listed first.

So we prefer to use fileEntries and directoryEntries to hold file and directory separately with alphabet order for this new API. 

Thanks,
Jiachen

File: lib/services/blob/blobservice.core.js (hunk 1 above)
Comment (ItalyPaleAle) [reply lib/services/blob/blobservice.core.js]: Yes, but I need some help. I looked at the test suite and it's very complex, and not sure how to integrate a new test.

File: lib/services/blob/blobservice.core.js [hunk 2]
```diff
@@ -1122,6 +1122,38 @@ BlobService.prototype.listBlobDirectoriesSegmentedWithPrefix = function (contain
...
+*                                                                 The maximum execution time interval begins at the time that the client begins building the request. The maximum
+*                                                                 execution time is checked intermittently while performing requests, and before executing retries.
+* @param {string}             [options.clientRequestId]           A string that represents the client request ID with a 1KB character limit.
+* @param {bool}               [options.useNagleAlgorithm]         Determines whether the Nagle algorithm is used; true to use the Nagle algorithm; otherwise, false.
+*                                                                 The default value is false.
+* @param {errorOrResult}      callback                            `error` will contain information
+*                                                                 if an error occurs; otherwise `result` will contain `entries` and `continuationToken`. 
+*                                                                 `entries`  gives a list of `[directories]{@link BlobResult}` and the `continuationToken` is used for the next listing operation.
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply lib/services/blob/blobservice.core.js]: Ok I'll look into that

**Comment by ItalyPaleAle** [comment]
Thanks for the comments, I'll make the changes you requested in the next couple of days.

With regards to tests: I can't really understand how the test suite works. I'm happy to contribute with a test, but would need some guidance on where to start.

**Comment by jiacfan** [comment]
@ItalyPaleAle 

Thanks for contribution, for testing, it's ok, we can discuss more on any question you met. And we can help to add tests as well. Please make sure the function work as your expectation in your side would be fine.

Thanks,
Jiachen

**Comment by ItalyPaleAle** [comment]
Question for you...

Currently the `_listBlobsOrDirectoriesSegmentedWithPrefix` method returns the results in `responseObject.listBlobsResult.entries`, for both directories and files.

1. If I implement the change as you request, then we'd have two different scenarios:
a. When `listBlobType` is `BlobConstants.ListBlobTypes.Directory` or `BlobConstants.ListBlobTypes.Blob`, the resulting list is in the `entries` key, whether it's files or directories. However, when `listBlobType` is `BlobConstants.ListBlobTypes.Any`, there would be two separate keys listing directories or files. This sounds a bit confusing to me.
2. The other option is to modify the method so directories are always returned in the `directoryEntries` key, and files are always returned in the `fileEntries` key, regardless of what `listBlobType` constant you pass. This would make the API simpler to understand and use (there are no different return keys depending on the arguments), but it's a backwards-incompatible change.

The risk of introducing backwards incompatibility should be lower considering that `_listBlobsOrDirectoriesSegmentedWithPrefix` is a private method, and it's always possible to remap everything to the `entries` key in the "wrapper methods" (e.g. `listBlobDirectoriesSegmentedWithPrefix`).

What do you think?

**Comment by ItalyPaleAle** [comment]
Actually there might be a third way. When passing any constant to listBlobType except Any, the method could return the same list both in the entries key and in one of fileEntries or directoryEntries. This would ensure backwards compatibility and introduce the new API; if the object is the same, then it shouldn’t use more memory.
When passing the Any constant, only the new keys are returned (so no entries key).

**Comment by jiacfan** [comment]
Hey, @ItalyPaleAle 

Yes, I see your ideas, and backward compatibility is important. At same time, for method `listBlobDirectoriesSegmented` or `listBlobsSegmented`, `entries` is reasonable as well. If everything is re-designed again, personally I prefer to the Option 2, i.e. always use directoryEntries and fileEntries. And currently, I think Option 2 + remap idea is clean and good. What's your favorite one?

At same time, do you think `All` would be better than `Any`, I prefer to All as the type for both blob and directory.

Thanks,
Jiachen

**Comment by ItalyPaleAle** [comment]
I've implemented the changes more in line with option 3, so it's fully backwards compatible but adds the new API too. Although the method `_listBlobsOrDirectoriesSegmentedWithPrefix` is private, I can't exclude that people are using it, and I'd rather not break any code in a minor version bump... Let me know what you think, I can revert back any time.

**Comment by ItalyPaleAle** [comment]
Speaking of backwards compatibility, I should also probably alias `_listBlobsOrDircotriesSegmentedWithPrefix` (with the typo) to the correct one, and maybe throw a warning?

File: lib/services/blob/blobservice.core.js [hunk 3]
```diff
@@ -5752,7 +5810,7 @@ BlobService.prototype._getBlobToStream = function (container, blob, writeStream,
 *                                                                 the entries of blobs and the continuation token for the next listing operation.
 *                                                                 `response` will contain information related to this operation.
 */
-BlobService.prototype._listBlobsOrDircotriesSegmentedWithPrefix = function (container, prefix, currentToken, listBlobType, optionsOrCallback, callback) {
+BlobService.prototype._listBlobsOrDirectoriesSegmentedWithPrefix = function (container, prefix, currentToken, listBlobType, optionsOrCallback, callback) {
```
Comment (jiacfan) [inline lib/services/blob/blobservice.core.js]: Ok, consider the backward compatibility of method with underline make sense, as it's not prevented being used by existing user,  please add comment for TODO: renaming, and we will rename the method when there is major version. At same time I'm ok for the solution, keeping entries only for backward compatibility consideration of method _listBlobsOrDircotriesSegmentedWithPrefix, and please add comment for TODO: removing the entries in _listBlobsOrDircotriesSegmentedWithPrefix.

As you know, both keeping fileEntries, directoryEntries and entries in same method could be ambiguous, so we can remove entries, and use Option3+mapping, in the future breaking change.

File: lib/services/blob/blobservice.core.js [hunk 4]
```diff
@@ -1177,12 +1236,11 @@ BlobService.prototype.listBlobsSegmented = function (container, currentToken, op
 * @param {bool}               [options.useNagleAlgorithm]         Determines whether the Nagle algorithm is used; true to use the Nagle algorithm; otherwise, false.
 *                                                                 The default value is false.
 * @param {errorOrResult}      callback                            `error` will contain information
-*                                                                 if an error occurs; otherwise `result` will contain
-*                                                                 the entries of `[blobs]{@link BlobResult}` and the continuation token for the next listing operation.
-*                                                                 `response` will contain information related to this operation.
-*/
+*                                                                 if an error occurs; otherwise `result` will contain `fileEntries` and `continuationToken`. 
+*                                                                 `fileEntries` (with an alias in `entries` for backwards compatibility) gives a list of `[blobs]{@link BlobResult}` and the `continuationToken` is used for the next listing operation.
```
Comment (jiacfan) [inline lib/services/blob/blobservice.core.js]: Please do not change comments here, and only expose entries as before.

File: lib/services/blob/blobservice.core.js [hunk 5]
```diff
@@ -1110,16 +1110,75 @@ BlobService.prototype.listBlobDirectoriesSegmented = function (container, curren
 * @param {bool}               [options.useNagleAlgorithm]         Determines whether the Nagle algorithm is used; true to use the Nagle algorithm; otherwise, false.
 *                                                                 The default value is false.
 * @param {errorOrResult}      callback                            `error` will contain information
-*                                                                 if an error occurs; otherwise `result` will contain `entries` and `continuationToken`. 
-*                                                                 `entries`  gives a list of `[directories]{@link BlobResult}` and the `continuationToken` is used for the next listing operation.
+*                                                                 if an error occurs; otherwise `result` will contain `directoryEntries` and `continuationToken`. 
+*                                                                 `directoryEntries` (with an alias in `entries` for backwards compatibility) gives a list of `[directories]{@link DirectoryResult}` and the `continuationToken` is used for the next listing operation.
```
Comment (jiacfan) [inline lib/services/blob/blobservice.core.js]: Please do not change comments here, and only expose entries as before.

File: lib/services/blob/blobservice.core.js [hunk 6]
```diff
@@ -1082,8 +1082,8 @@ BlobService.prototype.deleteContainerIfExists = function (container, optionsOrCa
 * @param {bool}               [options.useNagleAlgorithm]         Determines whether the Nagle algorithm is used; true to use the Nagle algorithm; otherwise, false.
 *                                                                 The default value is false.
 * @param {errorOrResult}      callback                            `error` will contain information
-*                                                                 if an error occurs; otherwise `result` will contain `entries` and `continuationToken`. 
-*                                                                 `entries`  gives a list of `[directories]{@link DirectoryResult}` and the `continuationToken` is used for the next listing operation.
+*                                                                 if an error occurs; otherwise `result` will contain `directoryEntries` and `continuationToken`. 
+*                                                                 `directoryEntries` (with an alias in `entries` for backwards compatibility) gives a list of `[directories]{@link DirectoryResult}` and the `continuationToken` is used for the next listing operation.
```
Comment (jiacfan) [inline lib/services/blob/blobservice.core.js]: Please do not change comments here, and only expose entries as before.

File: lib/services/blob/blobservice.core.js [hunk 7]
```diff
@@ -1145,8 +1204,8 @@ BlobService.prototype.listBlobDirectoriesSegmentedWithPrefix = function (contain
 * @param {bool}               [options.useNagleAlgorithm]         Determines whether the Nagle algorithm is used; true to use the Nagle algorithm; otherwise, false.
 *                                                                 The default value is false.
 * @param {errorOrResult}      callback                            `error` will contain information
-*                                                                 if an error occurs; otherwise `result` will contain `entries` and `continuationToken`. 
-*                                                                 `entries`  gives a list of `[blobs]{@link BlobResult}` and the `continuationToken` is used for the next listing operation.
+*                                                                 if an error occurs; otherwise `result` will contain `fileEntries` and `continuationToken`. 
+*                                                                 `fileEntries` (with an alias in `entries` for backwards compatibility) gives a list of `[blobs]{@link BlobResult}` and the `continuationToken` is used for the next listing operation.
```
Comment (jiacfan) [inline lib/services/blob/blobservice.core.js]: Please do not change comments here, and only expose entries as before.

**Comment by jiacfan** [comment]
@ItalyPaleAle 

Thanks for the updates, added comments in the PR. Please check the testing failures in CI as well, and try to fix it.

At same time, the idea of keeping backward compatibility is great!

Thanks,
Jiachen

**Comment by Frankyfrankly** [comment]
Any news on this? I'm also waiting for those changes.

**Comment by Frankyfrankly** [comment]
Any estimates when it will be ready? I also need those changes.

@jiacfan @ItalyPaleAle Would the 'azure-storage.d.ts' file be updated as well? It's still missing 'listBlobDirectoriesSegmented' and 'listBlobDirectoriesSegmentedWithPrefix' methods f.ex. despite they are present in 'blobservice.core.js'

**Comment by jiacfan** [comment]
Hi, @Frankyfrankly 

Thanks for the suggestion! @ItalyPaleAle please let me know if you need our help.

Thanks,
Jiachen

**Comment by Frankyfrankly** [comment]
So 'listBlobDirectoriesSegmented' and 'listBlobDirectoriesSegmentedWithPrefix' has been added to 'azure-storage.d.ts' in this PR #506. Dut this new method would still need to be declared, as well as the new return model whish is going to include ContinuationToken and both Blobs and BlobPrefixes.

**Comment by jiacfan** [comment]
Hi, @ItalyPaleAle 

Appreciate your existing contributions! And please let us know if you wish to continue the contribution, or you wish us to take the ownership?

Thanks,
Jiachen

**Comment by guymguym** [comment]
Hey all, any plans to rebase and merge this?
We are also waiting for it for https://github.com/noobaa/noobaa-core.
TIA

**Comment by EmmaZhu** [comment]
Hi @guymguym ,

We are deprecating this repro, currently there's no plan to release from this repository.
We would suggest to use @storage/storage-blob for blob operations.
The new SDK repo is here: https://github.com/Azure/azure-sdk-for-js

Thanks
Emma

**Pull request closed without merging**